kimchi/src/tests/foreign_field_mul.rs

```diff
@@ -0,0 +1,1409 @@
+    target_plookups: bool,
+    left_input: &BigUint,
+    right_input: &BigUint,
+    foreign_field_modulus: &BigUint,
+    invalidations: Vec<((usize, usize), G::ScalarField)>,
+) -> (CircuitGateResult<()>, [Vec<G::ScalarField>; COLUMNS])
+where
+    G::BaseField: PrimeField,
+    EFqSponge: Clone + FqSponge<G::BaseField, G, G::ScalarField>,
+    EFrSponge: FrSponge<G::ScalarField>,
+{
+    // Create foreign field multiplication gates
+    let (mut next_row, mut gates) = CircuitGate::<G::ScalarField>::create_foreign_field_mul(0);
+
+    // Compute multiplication witness
+    let (mut witness, external_checks) =
+        foreign_field_mul::witness::create(left_input, right_input, foreign_field_modulus);
+
+    // Optionally also add external gate checks to circuit
+    if external_gates {
+        // Layout for this test (just an example, circuit designer has complete flexibility, where to put the checks)
+        //      0-1  ForeignFieldMul
+        //      2-3  ForeignFieldAdd (result bound addition)
+        //      4-7  multi-range-check (left multiplicand)
+        //      8-11 multi-range-check (right multiplicand)
+        //     12-15 multi-range-check (product1_lo, product1_hi_0, carry1_lo)
+        //     16-19 multi-range-check (result range check)
+        //     20-23 multi-range-check (quotient range check)
+
+        // Bound addition for multiplication result
+        CircuitGate::extend_single_foreign_field_add(&mut gates, &mut next_row);
+        gates.connect_cell_pair((1, 0), (2, 0));
+        gates.connect_cell_pair((1, 1), (2, 1));
+        gates.connect_cell_pair((1, 2), (2, 2));
+        external_checks
+            .extend_witness_bound_addition(&mut witness, &foreign_field_modulus.to_field_limbs());
+
+        // Left input multi-range-check
+        CircuitGate::extend_multi_range_check(&mut gates, &mut next_row);
+        gates.connect_cell_pair((0, 0), (4, 0));
+        gates.connect_cell_pair((0, 1), (5, 0));
+        gates.connect_cell_pair((0, 2), (6, 0));
+        range_check::witness::extend_multi_limbs(&mut witness, &left_input.to_field_limbs());
+
+        // Right input multi-range-check
+        CircuitGate::extend_multi_range_check(&mut gates, &mut next_row);
+        gates.connect_cell_pair((0, 3), (8, 0));
+        gates.connect_cell_pair((0, 4), (9, 0));
+        gates.connect_cell_pair((0, 5), (10, 0));
+        range_check::witness::extend_multi_limbs(&mut witness, &right_input.to_field_limbs());
+
+        // Multiplication witness value product1_lo, product1_hi_0, carry1_lo multi-range-check
+        CircuitGate::extend_multi_range_check(&mut gates, &mut next_row);
+        gates.connect_cell_pair((0, 6), (12, 0)); // carry1_lo
+        gates.connect_cell_pair((1, 5), (13, 0)); // product1_lo
+        gates.connect_cell_pair((1, 6), (14, 0)); // product1_hi_0
+                                                  // Witness updated below
+
+        // Result/remainder bound multi-range-check
```

Comment:
Okay, seems okay to me, this is checking `result < fmod` when put together.

Reply:
right!